server/app/controllers/FileController.java

```diff
@@ -66,6 +73,44 @@ public CompletionStage<Result> show(Request request, long applicantId, String fi
 
   @Secure(authorizers = Authorizers.Labels.ANY_ADMIN)
   public Result adminShow(Request request, long programId, String fileKey) {
+    return adminShowInternal(request, Optional.of(programId), fileKey);
+  }
+
+  @Secure(authorizers = Authorizers.Labels.ANY_ADMIN)
+  public Result acledAdminShow(Request request, String fileKey) {
+    return adminShowInternal(request, /* maybeProgramId= */ Optional.empty(), fileKey);
+  }
+
+  public Result adminShowInternal(Request request, Optional<Long> maybeProgramId, String fileKey) {
+    Optional<StoredFile> maybeFile =
+        storedFileRepository.lookupFile(fileKey).toCompletableFuture().join();
+
+    if (maybeFile.isEmpty()) {
+      return notFound();
+    }
+
+    Account adminAccount =
+        profileUtils.currentUserProfile(request).orElseThrow().getAccount().join();
+
+    if (!maybeFile.get().getAcls().hasProgramReadPermission(adminAccount)) {
+      // If the request includes a program ID in the URL, try the legacy logic
+      // that assumes the admin is associated with the original program the
+      // file was uploaded for.
+      return maybeProgramId
+          .map(programId -> legacyAdminShow(request, programId, fileKey))
+          .orElse(unauthorized());
+    }
+
+    String decodedFileKey = URLDecoder.decode(fileKey, StandardCharsets.UTF_8);
+    return redirect(storageClient.getPresignedUrlString(decodedFileKey));
+  }
+
+  /**
+   * The old auth logic assumes that the admin viewing the file should be an admin for the original
+   * program the file was uploaded for. This is not the case, since file upload questions can be
+   * shared between programs.
+   */
+  private Result legacyAdminShow(Request request, long programId, String fileKey) {
```

Comment:
done

Reply:
I didn't change the file key format since it still serves it's purpose of creating a unique and readable file key for each uploaded file. The program ID embedded in it is useful for that, but we don't need to use it for auth logic anymore.

Reply:
done

Reply:
Ah ok, so likely we just need to give each file a unique (db sequence/id) value for uniqueness and not rely on the program construction that it was created under for that pseudo uniqueness.